MacVim won't build on El Capitan

### Steps to reproduce

Try to build MacVim from HEAD on El Cap like always.

### Expected behaviour

I expect it to build, like always.  Instead it bombs out in the GUI phase:
```
gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_MACVIM -Wall -Wno-unknown-pragmas -pipe  -DMACOS_X -DMACOS_X_DARWIN  -g -O2 -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1        -fobjc-exceptions -o objects/MMBackend.o MacVim/MMBackend.m
In file included from MacVim/gui_macvim.m:16:
In file included from MacVim/MMBackend.h:13:
In file included from ./vim.h:290:
./os_mac.h:304:5: error: unknown type name 'clockid_t'; did you mean 'clock_id_t'?
    clockid_t clockid,
    ^~~~~~~~~
    clock_id_t
/usr/include/mach/clock_types.h:53:13: note: 'clock_id_t' declared here
typedef int     clock_id_t;                     /* clock identification type */
                ^
1 error generated.
make[1]: *** [objects/gui_macvim.o] Error 1
```

### Version of Vim and architecture

HEAD

### Environment

MacOS El Capitan

### How MacVim was installed

clone the repo

### Logs and stack traces

_No response_

### Vim configuration where issue is reproducable

_No response_

### Issue has been tested with given configuration

- [ ] by running MacVim.app from GUI macOS interface
- [ ] by running vim/gvim/etc installed by MacVim
- [ ] by running other versions of vim (e.g. /usr/bin/vim)

### Issue has been tested with no configuration

- [ ] by running `mvim --clean` (or `gvim`, supplied by MacVim distribution)
- [ ] by running `vim --clean` (in terminal, supplied by MacVim distribution)
- [ ] by running `vim --clean` (in terminal, other suppliers, e.g. /usr/bin/vim)

### Other conditions

- [ ] The both Homebrew packages "vim" and "macvim" are installed

--- a/src/MacVim/MMFindReplaceController.m
+++ b/src/MacVim/MMFindReplaceController.m
@@ -8,6 +8,7 @@
  * See README.txt for an overview of the Vim source code.
  */
 
+#import "MacVim.h"
 #import "MMFindReplaceController.h"
 
 

--- a/src/MacVim/MMTextViewHelper.m
+++ b/src/MacVim/MMTextViewHelper.m
@@ -1072,7 +1072,7 @@ KeyboardInputSourcesEqual(TISInputSourceRef a, TISInputSourceRef b)
     static NSCursor *ibeamCursor = nil;
 
     if (!ibeamCursor) {
-        if (floor(NSAppKitVersionNumber) >= NSAppKitVersionNumber10_14)
+        if (AVAILABLE_MAC_OS(10, 14))
         {
             // macOS 10.14 (Mojave) introduced dark mode, and seems to have
             // added a thick white border around the system I-beam cursor,

--- a/src/MacVim/MMVimController.m
+++ b/src/MacVim/MMVimController.m
@@ -177,7 +177,7 @@ static BOOL isUnsafeMessage(int msgid);
     popupMenuItems = [[NSMutableArray alloc] init];
     toolbarItemDict = [[NSMutableDictionary alloc] init];
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12_2
-    if (@available(macos 10.12.2, *)) {
+    if (AVAILABLE_MAC_OS_PATCH(10, 12, 2)) {
         touchbarInfo = [[MMTouchBarInfo alloc] init];
     }
 #endif
@@ -1425,7 +1425,7 @@ static BOOL isUnsafeMessage(int msgid);
 
     if ([rootName isEqual:MMTouchbarMenuName]) {
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12_2
-        if (@available(macos 10.12.2, *)) {
+        if (AVAILABLE_MAC_OS_PATCH(10, 12, 2)) {
             if ([desc count] < 2) // Cannot be 1, as we need at least TouchBar.<menu_name>
                 return;
             if ([desc count] >= 3) // Unfortunately currently Apple does not support nested popover's so we can only do one level nesting
@@ -1509,7 +1509,7 @@ static BOOL isUnsafeMessage(int msgid);
         if ([desc count] >= 4) // Unfortunately currently Apple does not support nested popover's so we can only do one level nesting
             return;
 
-        if (@available(macos 10.12.2, *)) {
+        if (AVAILABLE_MAC_OS_PATCH(10, 12, 2)) {
             MMTouchBarInfo *submenuTouchbar = nil;
             if (![self touchBarItemForDescriptor:desc touchBar:&submenuTouchbar touchBarItem:nil]) {
                 return;
@@ -1597,7 +1597,7 @@ static BOOL isUnsafeMessage(int msgid);
     }
     if ([rootName isEqual:MMTouchbarMenuName]){
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12_2
-        if (@available(macos 10.12.2, *)) {
+        if (AVAILABLE_MAC_OS_PATCH(10, 12, 2)) {
             MMTouchBarInfo *submenuTouchbar = nil;
             if (![self touchBarItemForDescriptor:desc touchBar:&submenuTouchbar touchBarItem:nil]) {
                 return;
@@ -1651,7 +1651,7 @@ static BOOL isUnsafeMessage(int msgid);
 
     if ([rootName isEqual:MMTouchbarMenuName]) {
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12_2
-        if (@available(macos 10.12.2, *)) {
+        if (AVAILABLE_MAC_OS_PATCH(10, 12, 2)) {
             MMTouchBarItemInfo *touchbarItem = nil;
             if (![self touchBarItemForDescriptor:desc touchBar:nil touchBarItem:&touchbarItem]) {
                 return;
@@ -1695,7 +1695,7 @@ static BOOL isUnsafeMessage(int msgid);
 
     if ([rootName isEqual:MMTouchbarMenuName]) {
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_12_2
-        if (@available(macos 10.12.2, *)) {
+        if (AVAILABLE_MAC_OS_PATCH(10, 12, 2)) {
             MMTouchBarItemInfo *touchbarItem = nil;
             if (![self touchBarItemForDescriptor:desc touchBar:nil touchBarItem:&touchbarItem]) {
                 return;

--- a/src/MacVim/MMWindowController.m
+++ b/src/MacVim/MMWindowController.m
@@ -142,7 +142,7 @@
     // setting it in macOS 11+.
     BOOL usingTexturedBackground = NO;
 #if MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_VERSION_11_0
-    if (@available(macos 11.0, *)) {
+    if (AVAILABLE_MAC_OS(11, 0)) {
         // Don't set the textured background because it's been completely deprecated and won't do anything.
     } else {
         styleMask = styleMask | NSWindowStyleMaskTexturedBackground;
@@ -592,7 +592,7 @@
     
     // Transparent title bar setting
 #if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_10
-    if (@available(macos 10.10, *)) {
+    if (AVAILABLE_MAC_OS(10, 10)) {
         decoratedWindow.titlebarAppearsTransparent = [ud boolForKey:MMTitlebarAppearsTransparentKey];
     }
 #endif
@@ -1800,7 +1800,7 @@
     // See initWithVimController: for textured background deprecation notes.
     BOOL windowTextured = NO;
 #if MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_VERSION_11_0
-    if (@available(macos 11.0, *)) {
+    if (AVAILABLE_MAC_OS(11, 0)) {
     } else {
         windowTextured = ([decoratedWindow styleMask] &
                           NSWindowStyleMaskTexturedBackground) != 0;

--- a/src/MacVim/MacVim.h
+++ b/src/MacVim/MacVim.h
@@ -62,11 +62,33 @@
 #ifndef NSAppKitVersionNumber10_12
 # define NSAppKitVersionNumber10_12 1504
 #endif
+#ifndef NSAppKitVersionNumber10_12_2
+# define NSAppKitVersionNumber10_12_2 1504.76
+#endif
 #ifndef NSAppKitVersionNumber10_13
 # define NSAppKitVersionNumber10_13 1561
 #endif
 #ifndef NSAppKitVersionNumber10_14
 # define NSAppKitVersionNumber10_14 1671
+#endif
+#ifndef NSAppKitVersionNumber11_0
+# define NSAppKitVersionNumber11_0 2022
+#endif
+
+// Macro to detect runtime OS version. Ideally, we would just like to use
+// @available to test for this because the compiler can optimize it out
+// depending on your min/max OS configuration. However, it was added in Xcode 9
+// (macOS 10.13 SDK). For any code that we want to be compilable for Xcode 8
+// (macOS 10.12) or below, we need to use the macro below which will
+// selectively use NSAppKitVersionNumber instead.
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_13
+// Xcode 9+, can use @available, which is more efficient.
+# define AVAILABLE_MAC_OS(MAJOR, MINOR) @available(macos MAJOR##.##MINOR, *)
+# define AVAILABLE_MAC_OS_PATCH(MAJOR, MINOR, PATCH) @available(macos MAJOR##.##MINOR##.##PATCH, *)
+#else
+// Xcode 8 or below. Use the old-school NSAppKitVersionNumber check.
+# define AVAILABLE_MAC_OS(MAJOR, MINOR) NSAppKitVersionNumber >= NSAppKitVersionNumber##MAJOR##_##MINOR
+# define AVAILABLE_MAC_OS_PATCH(MAJOR, MINOR, PATCH) NSAppKitVersionNumber >= NSAppKitVersionNumber##MAJOR##_##MINOR##_##PATCH
 #endif
 
 // Deprecated constants. Since these are constants, we just need the compiler,
@@ -111,6 +133,10 @@
 // Deprecated constants in 10.13 SDK
 #define NSControlStateValueOn NSOnState
 #define NSControlStateValueOff NSOffState
+
+// Newly introduced symbols in 10.13 SDK
+typedef NSString* NSPasteboardType;
+typedef NSString* NSAttributedStringKey;
 #endif
 
 // Deprecated runtime values. Since these are runtime values, we need to use the

--- a/src/os_mac.h
+++ b/src/os_mac.h
@@ -272,7 +272,8 @@
 
 # include <dispatch/dispatch.h>
 
-# ifndef MAC_OS_X_VERSION_10_12
+# if !defined(MAC_OS_X_VERSION_10_12) || \
+	(MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_12)
 typedef int clockid_t;
 # endif
 # ifndef CLOCK_REALTIME

